src/main/java/io/divolte/server/filesinks/FileFlusher.java

```diff
@@ -0,0 +1,295 @@
+            logger.info("Recovered file system connection when creating file: {}", trackedFile);
+            return CONTINUE;
+        } catch (final IOException e) {
+            logger.error("File system connection error. Marking file system as unavailable. Attempting reconnect after "
+                    + reconnectDelay + " ms.", e);
+            markFileSystemUnavailable(timeMillis);
+            return PAUSE;
+        }
+    }
+
+    @Override
+    public void cleanup() {
+        currentTrackedFile.ifPresent(trackedFile -> {
+            try {
+                if (trackedFile.totalRecords + trackedFile.recordsSinceLastSync > 0) {
+                    trackedFile.divolteFile.closeAndPublish();
+                } else {
+                    trackedFile.divolteFile.discard();
+                }
+            } catch (final IOException ioe) {
+                logger.error("Failed to close and publish file " + trackedFile + " during cleanup.", ioe);
+            }
+        });
+    }
+
+    private void markFileSystemUnavailable(final long time) {
+        lastFixAttempt = time;
+        discardCurrentTrackedFileQuietly();
+        currentTrackedFile = Optional.empty();
+    }
+
+    private void possiblySyncAndOrRoll(final long time) throws IOException {
+        final TrackedFile trackedFile = currentTrackedFile.get();
+        if (time > trackedFile.projectedCloseTime) {
+            // roll file
+            if (trackedFile.totalRecords > 0) {
+                // Assumes closeAndPublish performs an implicit sync / flush of any remaining
+                // internal buffers. There no additional sync call here, as file manager
+                // implementations may have more optimal ways of performing a sync + close +
+                // move on the file in one go.
+                trackedFile.divolteFile.closeAndPublish();
+            } else {
+                trackedFile.divolteFile.discard();
+            }
+
+            currentTrackedFile = Optional.of(new TrackedFile(manager.createFile(newFileName())));
+        } else if (trackedFile.recordsSinceLastSync >= syncEveryRecords ||
+                time - trackedFile.lastSyncTime >= syncEveryMillis && trackedFile.recordsSinceLastSync > 0) {
+            // sync
+            sync(time, trackedFile);
+        } else if (trackedFile.recordsSinceLastSync == 0) {
+            // if nothing was written and we didn't roll, reset sync timing
+            trackedFile.lastSyncTime = time;
+        }
+    }
+
+    private void sync(final long time, final TrackedFile trackedFile) throws IOException {
+        trackedFile.divolteFile.sync();
+        trackedFile.totalRecords += trackedFile.recordsSinceLastSync;
+        trackedFile.recordsSinceLastSync = 0;
+        trackedFile.lastSyncTime = time;
+    }
+
+    private final class TrackedFile {
+        final long openTime;
+        final long projectedCloseTime;
+
+        final DivolteFile divolteFile;
+
+        long lastSyncTime;
```

Comment:
On top of that, under some circumstances syncing is evaluated but doesn't actually happen in absence of any pending records to be synced. I'm not sure if there's a better matching name for this field. Leaving as is.